### PR TITLE
Add compile_args to deep ensemble model

### PR DESCRIPTION
**Pull request by nfergu**
## Summary

This PR adds a `compile_args` option to the `DeepEnsemble` class. This enables various options, like JIT compilation for example, to be passed through to Keras's `compile` method.

**Fully backwards compatible:** yes

## PR checklist

- [] The quality checks are all passing
- [X] The bug case / new feature is covered by tests
- [X] Any new features are well-documented (in docstrings or notebooks)

All the tests in `tests/unit/models/keras` `tests/unit/models/gpflux` seem to pass locally and I have run `tox -e reformat` and `tox -e types` locally.

### CODE REVIEW

File: trieste/models/keras/models.py [hunk 1]
```diff
@@ -106,15 +107,28 @@ def __init__(
...
         """
         if model.ensemble_size < 2:
             raise ValueError(f"Ensemble size must be greater than 1 but got {model.ensemble_size}.")
 
         super().__init__(optimizer)
 
+        if compile_args is None:
+            compile_args = {}
+
+        if "optimizer" in compile_args or "loss" in compile_args or "metrics" in compile_args:
```
Comment (uri-granta) [inline trieste/models/keras/models.py]: (following is shorter, but possibly less clear?)

```suggestion
        if not {"optimizer", "loss", "metrics"}.isdisjoint(compile_args):
```

File: tests/unit/models/gpflux/test_models.py [hunk 2]
```diff
@@ -553,3 +553,22 @@ def test_deepgp_log(
...
+    with pytest.raises(
+        ValueError, match="You cannot enable `run_eagerly` and `jit_compile` at the same time."
+    ):
+        DeepGaussianProcess(model, compile_args={"jit_compile": True, "run_eagerly": True})
+
+
+def test_deepgp_disallowed_compile_args_specified() -> None:
+    mock_model = unittest.mock.MagicMock(spec=DeepGP)
+    with pytest.raises(ValueError):
+        DeepGaussianProcess(mock_model, compile_args={"optimizer": unittest.mock.MagicMock()})
```
Comment (uri-granta) [inline tests/unit/models/gpflux/test_models.py]: Test against non-trivial compile args maybe?

```suggestion
        DeepGaussianProcess(mock_model, compile_args={"run_eagerly": True, "optimizer": unittest.mock.MagicMock()})
```

**Review by uri-granta — APPROVED** [review]
LGTM! Tell me once you're happy and I'll merge it.

**Comment by nfergu** [comment]
> LGTM! Tell me once you're happy and I'll merge it.

@uri-granta Thanks. I committed your suggestions and applied them a bit more widely in [this commit](https://github.com/secondmind-labs/trieste/pull/816/commits/6d25215b895fca24b48cc70d0719b4bb13513620).

I'm happy with this, assuming all the checks pass now.

**Comment by nfergu** [comment]
@uri-granta I think you need to kick off the checks on this?

**Pull request merged**